File: python/open_redirect/checkout_flow.py
from flask import Flask, request, redirect

app = Flask(__name__)


# VULN 1: Open redirect after login - next parameter not validated
@app.route("/login", methods=["POST"])
def login():
    username = request.form.get("username")
    password = request.form.get("password")
    if authenticate(username, password):
        next_url = request.args.get("next", "/dashboard")
        return redirect(next_url)
    return "Login failed", 401


# VULN 2: Open redirect after checkout - return_url not validated against whitelist
@app.route("/checkout/complete")
def checkout_complete():
    return_url = request.args.get("return_url", "/orders")
    return redirect(return_url)


# VULN 3: Open redirect in OAuth callback - state parameter used as redirect target
@app.route("/oauth/callback")
def oauth_callback():
    code = request.args.get("code")
    state = request.args.get("state")
    # exchange code for token...
    return redirect(state)


def authenticate(username, password):
    return True

File: php/open_redirect/checkout_flow.php
<?php
/**
 * Shopist - Checkout Flow Redirects
 * DEMO FILE: Intentionally vulnerable for Datadog Code Security SAST demo.
 * DO NOT USE IN PRODUCTION.
 */

session_start();

// VULN 1: header() redirect to unsanitized $_GET['next'] after login (Open Redirect)
function handleLoginRedirect() {
    $username = $_POST['username'] ?? '';
    $password = $_POST['password'] ?? '';

    // Assume authentication succeeded
    $_SESSION['user'] = $username;

    // next parameter is attacker-controlled: ?next=https://evil.com/phish
    $next = $_GET['next'] ?? '/dashboard';
    header("Location: " . $next);
    exit();
}

// VULN 2: header() redirect to unsanitized $_POST['return_url'] after checkout (Open Redirect)
function completeCheckout() {
    $orderId = $_POST['order_id'] ?? 0;

    // Order processing logic ...
    $_SESSION['last_order'] = $orderId;

    // return_url from POST form field — attacker can set any external URL
    $returnUrl = $_POST['return_url'];
    header("Location: " . $returnUrl);
    exit();
}

// VULN 3: header() redirect to unsanitized $_GET['state'] in OAuth callback (Open Redirect)
function handleOAuthCallback() {
    $code  = $_GET['code']  ?? '';
    $state = $_GET['state'] ?? '/account';

    // OAuth code exchange logic ...
    // state parameter used as redirect target — should be validated against stored value
    header("Location: " . $state);
    exit();
}

// --- Route dispatcher ---
$action = $_GET['action'] ?? '';

if ($action === 'login') {
    handleLoginRedirect();
} elseif ($action === 'checkout_complete') {
    completeCheckout();
} elseif ($action === 'oauth_callback') {
    handleOAuthCallback();
}
